### ISSUE TITLE
Triton Error [CUDA]: an illegal memory access was encountered 

### ISSUE THREAD

**Issue by aifartist** [pytorch/pytorch#95312]
### 🐛 Describe the bug

`Triton Error [CUDA]: an illegal memory access was encountered`
I'm seeing this when trying to incorporate torch.compile into a largish app called AUTOMATIC1111 Stable Diffusion.  While I can't yet create a simple test case, I've done some debugging and will try to explain what I believe is happening in the code.  Forgive terminology errors as I'm not even an amateur yet.

To optimize, torch uses triton which creates triton.jit annotated functions to benchmark various options to determine which performs best.  In my case, there are 12 of these py files with the same function in each:
```
@triton.jit
def triton_mm(in_ptr0, arg_A, arg_B, out_ptr0):
    <body removed for brevity, full version in repro section>
    tl.store(out_ptr0 + (xindex + tl.zeros(mask.shape, tl.int32)), tmp1, mask)
```
I suspect xindex ends up with values that cause memory writes outside the boundaries of the out_ptr0 Tensor.  In fact, if I run the same function outside the real app with some random Tensor data, with the correct ?shape?, I can repro the same error, in my stand alone harness, by just changing the store() to something like "out_ptr0 **- 999999999** + (xindex..." to force a boundary violation.

Triton currently lacks the ability to print variables otherwise I would have just added a print of xindex to see if it has any bad offsets.  Using cuda-gdb does catch some problem inside of triton_mm but this is just a ?GPU? machine instruction with a one line stack trace.  And there is no debug info to print any variables in the debugger.

Today I will see if I can do something crazy like replace the tl.store above with:
```
    tl.store(out_ptr0, xindex)
```
I found where the real python call->run() site is within torch/_inductor/select_algorithm.py.  If it doesn't crash and the ?shapes? match I should be able to print the returned data in the real non triton jit code when the function returns.  I'm still learning python so maybe this won't work.  If I do find bad xindex values I can then work backwards in the function to see how they were computed.  Any advice would be helpful as I'm a python novice.  Turn around time on adding instrumentation and re-runing to a new crash is only about a minute.

@ptrblck 

### Error logs

Triton Error [CUDA]: an illegal memory access was encountered
```
CUDA Exception: Warp Out-of-range Address
The exception was triggered at PC 0x7f88d360eb90
Thread 91 "python3" received signal CUDA_EXCEPTION_5, Warp Out-of-range Address.
[Switching focus to CUDA kernel 0, grid 22591, block (0,0,0), thread (128,0,0), device 0, sm 0, warp 4, lane 0]
0x00007f88d360ebb0 in triton_mm_0d1d2d3d<<<(40,1,1),(256,1,1)>>> ()
```
There is just the one line in the stack trace.
x/i $pc gives me:

```
    HMMA.16816.F32 R4, R12, R8, R4
```
```
R4 = 0x411a7319
R8 = 0x3943bb2d
R12 = 0x305439af
```

### Minified repro

Here is the version that crashes.  The other 11 which don't crash have EVEN_K = True.  This isn't a repro but someone might have an idea as to what and how to instrument this for me to debug it.
```
import triton.language as tl
import triton
from torch._inductor.triton_ops.autotune import template
from torch._inductor.utils import instance_descriptor

@triton.jit
def triton_mm(in_ptr0, arg_A, arg_B, out_ptr0):
    GROUP_M : tl.constexpr = 8
    EVEN_K : tl.constexpr = False
    ALLOW_TF32 : tl.constexpr = True
    ACC_TYPE : tl.constexpr = tl.float32
    BLOCK_M : tl.constexpr = 32
    BLOCK_N : tl.constexpr = 32
    BLOCK_K : tl.constexpr = 128

    A = arg_A
    B = arg_B

    M = 2
    N = 1280
    K = 320
    stride_am = 320
    stride_ak = 1
    stride_bk = 1
    stride_bn = 320

    # based on triton.ops.matmul
    pid = tl.program_id(0)
    grid_m = (M + BLOCK_M - 1) // BLOCK_M
    grid_n = (N + BLOCK_N - 1) // BLOCK_N

    # re-order program ID for better L2 performance
    width = GROUP_M * grid_n
    group_id = pid // width
    group_size = min(grid_m - group_id * GROUP_M, GROUP_M)
    pid_m = group_id * GROUP_M + (pid % group_size)
    pid_n = (pid % width) // (group_size)

    rm = pid_m * BLOCK_M + tl.arange(0, BLOCK_M)
    rn = pid_n * BLOCK_N + tl.arange(0, BLOCK_N)
    ram = tl.max_contiguous(tl.multiple_of(rm % M, BLOCK_M), BLOCK_M)
    rbn = tl.max_contiguous(tl.multiple_of(rn % N, BLOCK_N), BLOCK_N)
    rk = tl.arange(0, BLOCK_K)
    A = A + (ram[:, None] * stride_am + rk[None, :] * stride_ak)
    B = B + (rk[:, None] * stride_bk + rbn[None, :] * stride_bn)

    acc = tl.zeros((BLOCK_M, BLOCK_N), dtype=ACC_TYPE)
    for k in range(K, 0, -BLOCK_K):
        if EVEN_K:
            a = tl.load(A)
            b = tl.load(B)
        else:
            a = tl.load(A, mask=rk[None, :] < k, other=0.)
            b = tl.load(B, mask=rk[:, None] < k, other=0.)
        acc += tl.dot(a, b, allow_tf32=ALLOW_TF32)
        A += BLOCK_K * stride_ak
        B += BLOCK_K * stride_bk

    # rematerialize rm and rn to save registers
    rm = pid_m * BLOCK_M + tl.arange(0, BLOCK_M)
    rn = pid_n * BLOCK_N + tl.arange(0, BLOCK_N)
    idx_m = rm[:, None]
    idx_n = rn[None, :]
    mask = (idx_m < M) & (idx_n < N)

    # inductor generates a suffix
    xindex = idx_n + (1280*idx_m)
    tmp0 = tl.load(in_ptr0 + (idx_n + tl.zeros(mask.shape, tl.int32)), mask).to(tl.float32)
    tmp1 = acc + tmp0
    tl.store(out_ptr0 + (xindex + tl.zeros(mask.shape, tl.int32)), tmp1, mask)
```
                                                                                                                                                                             

### Versions

Collecting environment information...
PyTorch version: 2.0.0.dev20230220+cu118
Is debug build: False
CUDA used to build PyTorch: 11.8
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.2 LTS (x86_64)
GCC version: (Ubuntu 11.3.0-1ubuntu1~22.04) 11.3.0
Clang version: 14.0.0-1ubuntu1
CMake version: version 3.25.2
Libc version: glibc-2.35

Python version: 3.10.6 (main, Nov 14 2022, 16:10:14) [GCC 11.3.0] (64-bit runtime)
Python platform: Linux-5.19.0-32-generic-x86_64-with-glibc2.35
Is CUDA available: True
CUDA runtime version: 11.8.89
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: GPU 0: NVIDIA GeForce RTX 4090
Nvidia driver version: 525.89.02
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.8.8.0
/usr/lib/x86_64-linux-gnu/libcudnn_adv_infer.so.8.8.0
/usr/lib/x86_64-linux-gnu/libcudnn_adv_train.so.8.8.0
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_infer.so.8.8.0
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_train.so.8.8.0
/usr/lib/x86_64-linux-gnu/libcudnn_ops_infer.so.8.8.0
/usr/lib/x86_64-linux-gnu/libcudnn_ops_train.so.8.8.0
/usr/local/cuda-12.0/targets/x86_64-linux/lib/libcudnn.so.8.7.0
/usr/local/cuda-12.0/targets/x86_64-linux/lib/libcudnn_adv_infer.so.8.7.0
/usr/local/cuda-12.0/targets/x86_64-linux/lib/libcudnn_adv_train.so.8.7.0
/usr/local/cuda-12.0/targets/x86_64-linux/lib/libcudnn_cnn_infer.so.8.7.0
/usr/local/cuda-12.0/targets/x86_64-linux/lib/libcudnn_cnn_train.so.8.7.0
/usr/local/cuda-12.0/targets/x86_64-linux/lib/libcudnn_ops_infer.so.8.7.0
/usr/local/cuda-12.0/targets/x86_64-linux/lib/libcudnn_ops_train.so.8.7.0
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                    x86_64
CPU op-mode(s):                  32-bit, 64-bit
Address sizes:                   46 bits physical, 48 bits virtual
Byte Order:                      Little Endian
CPU(s):                          32
On-line CPU(s) list:             0-31
Vendor ID:                       GenuineIntel
Model name:                      13th Gen Intel(R) Core(TM) i9-13900K
CPU family:                      6
Model:                           183
Thread(s) per core:              2
Core(s) per socket:              24
Socket(s):                       1
Stepping:                        1
CPU max MHz:                     5800.0000
CPU min MHz:                     800.0000
BogoMIPS:                        5990.40
Flags:                           fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf tsc_known_freq pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb invpcid_single ssbd ibrs ibpb stibp ibrs_enhanced tpr_shadow vnmi flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid rdseed adx smap clflushopt clwb intel_pt sha_ni xsaveopt xsavec xgetbv1 xsaves split_lock_detect avx_vnni dtherm ida arat pln pts hwp hwp_notify hwp_act_window hwp_epp hwp_pkg_req hfi umip pku ospke waitpkg gfni vaes vpclmulqdq tme rdpid movdiri movdir64b fsrm md_clear serialize pconfig arch_lbr ibt flush_l1d arch_capabilities
Virtualization:                  VT-x
L1d cache:                       896 KiB (24 instances)
L1i cache:                       1.3 MiB (24 instances)
L2 cache:                        32 MiB (12 instances)
L3 cache:                        36 MiB (1 instance)
NUMA node(s):                    1
NUMA node0 CPU(s):               0-31
Vulnerability Itlb multihit:     Not affected
Vulnerability L1tf:              Not affected
Vulnerability Mds:               Not affected
Vulnerability Meltdown:          Not affected
Vulnerability Mmio stale data:   Not affected
Vulnerability Retbleed:          Not affected
Vulnerability Spec store bypass: Mitigation; Speculative Store Bypass disabled via prctl
Vulnerability Spectre v1:        Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:        Mitigation; Enhanced IBRS, IBPB conditional, RSB filling, PBRSB-eIBRS SW sequence
Vulnerability Srbds:             Not affected
Vulnerability Tsx async abort:   Not affected

Versions of relevant libraries:
[pip3] mypy-extensions==1.0.0
[pip3] numpy==1.23.3
[pip3] open-clip-torch==2.7.0
[pip3] pytorch-lightning==1.7.6
[pip3] pytorch-triton==2.0.0+c8bfe3f548
[pip3] torch==2.0.0.dev20230220+cu118
[pip3] torchaudio==2.0.0.dev20230220+cu118
[pip3] torchdiffeq==0.2.3
[pip3] torchmetrics==0.11.1
[pip3] torchsde==0.2.5
[pip3] torchtriton==2.0.0+f16138d447
[pip3] torchvision==0.15.0.dev20230220+cu118
[conda] Could not collect

cc @ezyang @soumith @msaroufim @wconstab @ngimel @bdhirsh

**Comment by ezyang** [comment]
obligatory "Are you using the new triton MLIR backend?"

**Comment by ngimel** [comment]
I can repro on MLIR with 
```
import torch
from torch._inductor import debug
debug=True

def fn(inp, a, b):
    return torch.addmm(inp, a, b.t())

a=torch.randn(2, 320, device="cuda", dtype=torch.half)
b=torch.randn(1280, 320, device="cuda", dtype=torch.half)
out=torch.empty(1280, device="cuda", dtype=torch.half)
print(fn(out, a, b))
opt=torch.compile(fn, mode="max-autotune")
opt(out, a, b)
```

**Comment by aifartist** [comment]
> obligatory "Are you using the new triton MLIR backend?"

How do I do that?  Is that a suggested cause or a suggested solution?

**Comment by aifartist** [comment]
> I can repro on MLIR with
> 
> ```
> import torch
> from torch._inductor import debug
> debug=True
> 
> def fn(inp, a, b):
>     return torch.addmm(inp, a, b.t())
> 
> a=torch.randn(2, 320, device="cuda", dtype=torch.half)
> b=torch.randn(1280, 320, device="cuda", dtype=torch.half)
> out=torch.empty(1280, device="cuda", dtype=torch.half)
> print(fn(out, a, b))
> opt=torch.compile(fn, mode="max-autotune")
> opt(out, a, b)
> ```
Looks like this works as a repro.  It also repro's in my env.
FYI, I also get: 'sm_89' is not a recognized processor for this target (ignoring processor)
but I've been seeing that for awhile now.

**Comment by aifartist** [comment]
@ezyang @ngimel Is this a known issue with any workaround?

I have to use Triton 2 because I have a 4090 which isn't recognized if I don't.

**Comment by ngimel** [comment]
standalone triton repo https://gist.github.com/ngimel/88e408d9e7be6644788096a64ef5f301, `num_warps=8` doesn't work, `num_warps=4` does.

**Comment by ngimel** [comment]
>FYI, I also get: 'sm_89' is not a recognized processor for this target (ignoring processor)
but I've been seeing that for awhile now.

Hopefully this should be fixed in the next nightly

**Comment by parthmannan** [comment]
Is there a way to enable/disable the new MLIR backend to test if the error gets fixed using that?

**Comment by ezyang** [comment]
essentially, if you are on a new enough nightly, you should be getting the mlir backend with our triton packaging

**Comment by aifartist** [comment]
@ngimel @ezyang It's ALIVE!
After applying the fix provided and then finding another bug I fixed locally in guards.py:LIST_LENGTH(self, guard) it then appeared hung but I knew it was grinding away optimizing and once finished I got some nice perf numbers.  No great improvement but 43.5 it/s is quite fast compared with what I saw before.  More importantly I saw an average of only about 88% GPU usage on my 4090 which is lower than normal so more head room is available.  I've posted elsewhere the as the GPU and algorithms get faster the need for a fast CPU increases.  I have a i9-13900K which is as fast as they get for a single thread which is all that is used.

Now I want to see if using "@torch.compile" on a few key functions can speed up the CPU side so that I can push the 4090 faster.  There are bugs there also but I have fixes for them.

Thanks so much for the quick turn around.  I want to see if I can get torch.compile to work with backend/engine = tensorrt.

**Comment by parthmannan** [comment]
@aifartist Not sure if it will help but I would also try using cudagraphs to reduce CPU side bottlenecks. I add this before my `torch.compile` call. Worth a shot.

```
from torch._inductor import config as inductor_config
inductor_config.triton.cudagraphs = True
```

**Comment by Chillee** [comment]
@parthmannan You can also do `torch.compile(mode="reduce-overhead")`

**Comment by parthmannan** [comment]
Oh that's neat. I should go look up all the mode choices.
Can the mode arg take a list - such as `max-autotune` and `reduce-overhead` together?

**Comment by ngimel** [comment]
You can use `options` arg to torch.compile to pass all the options that you want https://pytorch.org/docs/master/generated/torch.compile.html?highlight=torch+compile#torch.compile

**Comment by aifartist** [comment]
@parthmannan I doubt if it will help considering I'm already using max-autotune and I just found the following in the source
so it appears cudagraphs get's used.  What I want to try now is to see if backend="tensorrt" works yet.  It is supposed to be the biggie in terms of perf improvement.  I think I need to manually build PyTorch to get USE_TENSORRT=1 support.
```
        elif mode == "reduce-overhead":
            self.apply_options({
                "triton.cudagraphs": True,
                "size_asserts": False,
            })
        elif mode == "max-autotune":
            self.apply_options({
                "epilogue_fusion": True,
                "max_autotune": True,
                "triton.cudagraphs": True,
            })
```

**Comment by aifartist** [comment]
> You can use `options` arg to torch.compile to pass all the options that you want https://pytorch.org/docs/master/generated/torch.compile.html?highlight=torch+compile#torch.compile

mode=default or mode=reduce-overhead will cause max-autotune to be ingored given the way the code is written.

**Comment by aifartist** [comment]
@ngimel As I suspected...
Given I was no longer pushing the GPU past 88% I tried a compile for batchsize=4 and went from 13 it/s to 15 it/s or a good 15% improvement.  An aggregate of 60 it/s.  Furthermore when I switch back to batchsize=1 without re compiling I my 43.5 went to 44.5 or more.  Obviously optimization will improve in the future.

I don't mind the 15 minutes to compile the model ONCE but now it seem to take about 1 minute to start each batch before the image generations even start if I restart the app.
